blockchain/blockchain.go

```diff
@@ -41,22 +42,66 @@ func DecodeBlockChain(r io.Reader) *BlockChain {
 	return &bc
 }
 
-// ValidTransaction checks whether a transaction is valid, assuming the blockchain is valid.
+// ValidTransaction checks whether a transaction is valid, assuming the
+// blockchain is valid.
 func (bc *BlockChain) ValidTransaction(t *Transaction) bool {
+
 	// Find the transaction input (I) in the chain (by hash)
+	var I *Transaction
+	inputBlock := bc.Blocks[t.Input.BlockNumber]
+	for _, transaction := range inputBlock.Transactions {
+		if t.Input.Hash == HashSum(transaction) {
+			I = transaction
+		}
+	}
+	if I == nil {
+		return false
+	}
+
 	// Check that output to sender in I is equal to outputs in T
+	var inAmount uint64
+	for _, output := range I.Outputs {
+		if output.Recipient == t.Sender {
+			inAmount += output.Amount
+		}
+	}
+	var outAmount uint64
+	for _, output := range t.Outputs {
+		outAmount += output.Amount
+	}
+	if inAmount != outAmount {
+		return false
+	}
+
```

Comment:
We also need to check that input transaction hasn't been spent between when it was created and now, right?